encode `gh browse` output url

bchadwic opened the pull request:
fixes #4647


1. User's before, and with this pr can't open up files with a `\` in the file's / folder's name
	- I can update the pr to allow for `\` to be escaped with another `\`? Not sure if this would be useful. `/` can't be in a path name so this isn't a problem.
2. I removed `utils.DisplayUrl()`, but this can be reinserted easily. 
	- I'm not sure if users would rather see the encoded or nonencoded url when opening up the url in the browser. Although I'm starting to see why that function was there in the first place.

Commit:
encoded special characters in browse url

bchadwic commented on pkg/cmd/browse/browse.go:
```diff
@@ -131,15 +133,19 @@ func runBrowse(opts *BrowseOptions) error {
 		return err
 	}
-	url := ghrepo.GenerateRepoURL(baseRepo, section)
+	// taken from cli/internal/ghrepo/repo.go
+	url := fmt.Sprintf("%s%s/%s", ghinstance.HostPrefix(baseRepo.RepoHost()), baseRepo.RepoOwner(), baseRepo.RepoName())
+	if section != "" {
+		url = fmt.Sprintf("%s/%s", url, section)
+	}
```
Also, just to be clear, I changed this because the encoded characters act like format specifiers for the `fmt.Sprintf(p, args...)` in `GenerateRepoURL()`. I copied the code as a simple hack around it?

mislav replied:
You can keep using GenerateRepoURL, but like this so to avoid issues with `%`:
```
GenerateRepoURL(baseRepo, "%s", section)
```
Note that `section` should already be path-escaped here

mislav commented on pkg/cmd/browse/browse.go:
```diff
@@ -131,15 +133,19 @@ func runBrowse(opts *BrowseOptions) error {
 	}
 
 	if opts.IO.IsStdoutTTY() {
-		fmt.Fprintf(opts.IO.Out, "Opening %s in your browser.\n", utils.DisplayURL(url))
```
Please keep using `DisplayURL` here

mislav commented on pkg/cmd/browse/browse.go:
```diff
@@ -201,8 +207,8 @@ func parseFile(opts BrowseOptions, f string) (p string, start int, end int, err
-		// Ensure that a path using \ can be used in a URL
-		p = strings.ReplaceAll(p, "\\", "/")
+		// encode special characters and replace all encoded slashes with forward slashes
+		p = regexp.MustCompile(`%2F|%5C`).ReplaceAllString(url.PathEscape(p), "/")
```
Let's only replace `%2F` back to `/`, but leave `%5C` intact.

mislav commented on pkg/cmd/browse/browse.go:
```diff
@@ -201,8 +207,8 @@ func parseFile(opts BrowseOptions, f string) (p string, start int, end int, err
 	if !filepath.IsAbs(p) {
 		p = filepath.Clean(filepath.Join(opts.PathFromRepoRoot(), p))
-		// Ensure that a path using \ can be used in a URL
-		p = strings.ReplaceAll(p, "\\", "/")
```
Let's not replace backslashes with forward slashes. Let backslashes be encoded as `%5c`.

mislav commented on pkg/cmd/browse/browse.go:
```diff
@@ -201,8 +207,8 @@ func parseFile(opts BrowseOptions, f string) (p string, start int, end int, err
 	p = parts[0]
 	if !filepath.IsAbs(p) {
 		p = filepath.Clean(filepath.Join(opts.PathFromRepoRoot(), p))
-		// Ensure that a path using \ can be used in a URL
```
It's not the responsibility of a function called `parseFile` to be escaping the resulting string for use in a URL later. You're on the right tracking with escaping, but it should be moved elsewhere, e.g. to just before this value gets used in a URL.

bchadwic committed:
updating changes requested

bchadwic commented:
@mislav 

Sorry, I'm a little lost on what to do with `\` according to the review. It sounds like we should keep `\` escaped after we perform `PathEscape`, does this pose problems for windows users? 

Files, on GitHub, with `\` in the name won't be able to be opened from `gh` if we decide  _not_ to escape them. If we do escape them, doesn't that mean this bug fix would cause a breaking change for windows users since currently `\` is an acceptable path separator on `gh`? I think I may have missed something.

mislav committed:
:nail_care: tweak how `browse` handles file paths

- URLs are now always generated without a trailing slash
- Windows-style paths are normalized before processing
- Special characters in branch names are escaped

mislav reviewed (approved):
Sorry, I got a bit lost in our conversation because I didn't have full context about what the current behavior is of handling Windows-style paths. I've pushed some changes that normalize paths on Windows (backslash separator is converted to slash) before further processing. This way, we don't have to worry about path separators on Windows being converted to URL escape sequences. I've also made it so that special characters in branch names are escaped too ✌️

mislav committed:
Merge remote-tracking branch 'origin' into browse-path-escape-patch

mislav committed:
Fix tests and ensure they don't write to stderr
